File: python/src/workshop_api/fitness/managing_memberships/suspend_membership/suspend_membership_controller.py
from datetime import UTC, datetime

from fastapi import APIRouter, Body, Depends
from sqlalchemy.orm import Session

from workshop_api.fitness.database import get_db_session

from ..shared.membership_entity import (
    MembershipBillingReferenceOrmModel,
    MembershipOrmModel,
)
from .suspend_overdue_memberships_request import (
    SuspendOverdueMembershipsRequest,
)
from .suspend_overdue_memberships_response import (
    SuspendOverdueMembershipsResponse,
)

router = APIRouter(prefix="/api/memberships", tags=["membership"])


@router.post(
    "/suspend-overdue",
    response_model=SuspendOverdueMembershipsResponse,
    response_model_by_alias=True,
)
def suspend_overdue_memberships(
    request: SuspendOverdueMembershipsRequest | None = Body(default=None),
    session: Session = Depends(get_db_session),
) -> SuspendOverdueMembershipsResponse:
    checked_at = request.checked_at if request and request.checked_at else datetime.now(UTC)
    checked_at_date = checked_at.astimezone(UTC).date() if checked_at.tzinfo else checked_at.date()

    memberships = session.query(MembershipOrmModel).all()
    open_billing_references = (
        session.query(MembershipBillingReferenceOrmModel)
        .filter(MembershipBillingReferenceOrmModel.status == "OPEN")
        .all()
    )

    suspended_membership_ids = []
    checked_memberships = 0
    for membership in memberships:
        if not membership.is_active():
            continue

        checked_memberships += 1
        overdue_billing_reference = next(
            (
                billing_reference
                for billing_reference in open_billing_references
                if billing_reference.membership_id == membership.id
                and billing_reference.due_date < checked_at_date
            ),
            None,
        )
        if overdue_billing_reference is None:
            continue

        if not membership.is_suspended_for_non_payment():
            membership.suspend_for_non_payment()
            suspended_membership_ids.append(membership.id)

    session.commit()

    return SuspendOverdueMembershipsResponse(
        checkedAt=checked_at,
        checkedMemberships=checked_memberships,
        suspendedMembershipIds=suspended_membership_ids,
    )

File: java/src/main/java/com/workshop/architecture/fitness/managing_memberships/suspend_membership/SuspendMembershipController.java
package com.workshop.architecture.fitness.managing_memberships.suspend_membership;

import com.workshop.architecture.fitness.managing_memberships.shared.MembershipBillingReferenceEntity;
import com.workshop.architecture.fitness.managing_memberships.shared.MembershipBillingReferenceRepository;
import com.workshop.architecture.fitness.managing_memberships.shared.MembershipEntity;
import com.workshop.architecture.fitness.managing_memberships.shared.MembershipRepository;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.List;

@RestController
@RequestMapping("/api/memberships")
public class SuspendMembershipController {

    private final MembershipRepository membershipRepository;
    private final MembershipBillingReferenceRepository billingReferenceRepository;

    public SuspendMembershipController(
            MembershipRepository membershipRepository,
            MembershipBillingReferenceRepository billingReferenceRepository
    ) {
        this.membershipRepository = membershipRepository;
        this.billingReferenceRepository = billingReferenceRepository;
    }

    @PostMapping("/suspend-overdue")
    SuspendOverdueMembershipsResponse suspendOverdueMemberships(
            @RequestBody(required = false) SuspendOverdueMembershipsRequest request
    ) {
        Instant checkedAt = request == null || request.checkedAt() == null ? Instant.now() : request.checkedAt();
        LocalDate checkedAtDate = checkedAt.atZone(ZoneOffset.UTC).toLocalDate();
        List<MembershipEntity> memberships = membershipRepository.findAll();
        List<MembershipBillingReferenceEntity> openBillingReferences = billingReferenceRepository.findByStatus("OPEN");
        List<String> suspendedMembershipIds = new ArrayList<>();
        int checkedMemberships = 0;

        for (MembershipEntity membership : memberships) {
            if (!"ACTIVE".equals(membership.getStatus())) {
                continue;
            }

            checkedMemberships++;

            MembershipBillingReferenceEntity overdueBillingReference = openBillingReferences.stream()
                    .filter(billingReference -> membership.getId().equals(billingReference.getMembershipId()))
                    .filter(billingReference -> billingReference.getDueDate().isBefore(checkedAtDate))
                    .findFirst()
                    .orElse(null);

            if (overdueBillingReference == null) {
                continue;
            }

            if (!membership.isSuspendedForNonPayment()) {
                membership.suspendForNonPayment();
                membershipRepository.save(membership);
                suspendedMembershipIds.add(membership.getId().toString());
            }
        }

        return new SuspendOverdueMembershipsResponse(
                checkedAt,
                checkedMemberships,
                suspendedMembershipIds
        );
    }

}
